### PR TITLE
Add spoonacular recipe api integration

### PR DESCRIPTION
**Pull request by mplekunov**

### CODE REVIEW

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts [hunk 1]
```diff
@@ -0,0 +1,312 @@
+import axios from 'axios';
+import * as dotenv from 'dotenv';
+import { URLSearchParams } from 'url';
+import IncorrectIDFormat from '../../exceptions/IncorrectIDFormat';
+import NoParameterFound from '../../exceptions/NoParameterFound';
+import ParameterIsNotAllowed from '../../exceptions/ParameterIsNotAllowed';
+import IFoodAPI from '../../foodAPI/IFoodAPI';
+import IFood from '../../serverAPI/model/food/IFood';
+import IInstruction from '../../serverAPI/model/instruction/IInstruction';
+dotenv.config();
```
Comment (nashirj) [inline server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:10]: Can this be after all the imports? Looks weird being in the middle.

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+            'cuisine',
+            'diet',
+            'intolerences',
+            'includeIngredients',
+            'type'
+        ]);
+
+        this.recipeGetParameters = new Set([
+
+        ]);
```
Comment (nashirj) [inline server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:59]: Is this needed?

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+
+    protected headers: any;
+
+    constructor(foodAPI: IFoodAPI) {
+        this.foodAPI = foodAPI;
+
+        this.apiKey = process.env.SPOONACULAR_API_KEY;
+        this.host = process.env.SPOONACULAR_HOST;
+
+        this.dishTypes = new Set([
```
Comment (nashirj) [inline server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:34]: Are these taken from spoonacular directly? If so, can you add a comment and a link to their docs saying so?

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+            "salad",
+            "bread",
+            "breakfast",
+            "soup",
+            "beverage",
+            "sauce",
+            "drink"
+        ]);
+
+        this.recipeSearchParameters = new Set([
```
Comment (nashirj) [inline server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:48]: Same comment about linking to spoonacular docs

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+            'query',
```
Comment (nashirj) [inline server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:49]: `recipeName`?

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+            'cuisine',
+            'diet',
+            'intolerences',
```
Comment (nashirj) [inline server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:52]: typo :(

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+        this.recipeGetParameters = new Set([
+
+        ]);
+    }
+
+    /**
+     * Retrieves collection of partially filled Recipe items that satisfy searching parameters.
+     * 
+     * @param parameters parameters used for searching.
+     * - query - required parameter that defines the name of the Recipe item (partial names are accepted).
```
Comment (nashirj) [inline server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:66]: `recipeName`

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+     * - number - optional parameter that defines max number of the results to be returned.
+     * - page - optional parameter that defines page number for pagination.
+     * - intolerence - optional parameter that defines the type of intolerences to be taken in consideration during searching.
+     * - cuisine - optional parameter that limits search results to specific cuisines.
+     * - type - optional parameter that limits search results to specific dish types.
+     * 
+     * @throws NoParameterFound exception when required parameters weren't found.
+     * @throws ParameterIsNotAllowed exception when encountering a non-existing parameter.
+     * 
+     * @returns Promise filled with a collection of Partial<IRecipe> objects.
```
Comment (nashirj) [inline server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:76]: Spacing

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+    }
+
+    /**
+     * Retrieves collection of partially filled Recipe items that satisfy searching parameters.
+     * 
+     * @param parameters parameters used for searching.
+     * - query - required parameter that defines the name of the Recipe item (partial names are accepted).
+     * - number - optional parameter that defines max number of the results to be returned.
+     * - page - optional parameter that defines page number for pagination.
+     * - intolerence - optional parameter that defines the type of intolerences to be taken in consideration during searching.
```
Comment (nashirj) [inline server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:69]: Please ctrl+f and replace all intolerence to intolerance everywhere in all your branches

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+     * - cuisine - optional parameter that limits search results to specific cuisines.
+     * - type - optional parameter that limits search results to specific dish types.
+     * 
+     * @throws NoParameterFound exception when required parameters weren't found.
+     * @throws ParameterIsNotAllowed exception when encountering a non-existing parameter.
```
Comment (nashirj) [inline server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:74]: `... when an invalid parameter is provided in the request`

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+
+        ]);
+    }
+
+    /**
+     * Retrieves collection of partially filled Recipe items that satisfy searching parameters.
+     * 
+     * @param parameters parameters used for searching.
+     * - query - required parameter that defines the name of the Recipe item (partial names are accepted).
+     * - number - optional parameter that defines max number of the results to be returned.
```
Comment (nashirj) [inline server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:67]: Please specify default values for optional parameters

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+        }
+
+        return recipeArray;
+    }
+
+    /**
+     * Converts search recipe parameters to URLSearchParams.
+     * 
+     * @param parameters parameters used for searching.
+     * - query - required parameter that defines the name of the Recipe item (partial names are accepted).
```
Comment (nashirj) [inline server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:107]: `recipeName`

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+     * - number - optional parameter that defines max number of the results to be returned.
+     * - page - optional parameter that defines page number for pagination.
+     * - intolerence - optional parameter that defines the type of intolerences to be taken in consideration during searching.
+     * - cuisine - optional parameter that limits search results to specific cuisines.
+     * - type - optional parameter that limits search results to specific dish types.
+     * 
+     * @throws NoParameterFound exception when required parameters weren't found.
+     * @throws ParameterIsNotAllowed exception when encountering a non-existing parameter.
```
Comment (nashirj) [inline server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:115]: Same as above

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+            instructionSteps: await this.parseInstructionSteps(recipeObject),
+            servings: recipeObject.servings,
+            preparationInMinutes: recipeObject.preparationMinutes,
+            cookingTimeInMinutes: recipeObject.readyInMinutes,
+            totalCost: recipeObject.pricePerServing * recipeObject.servings
+        }
+    }
+
+    protected async parseInstructionSteps(recipeObject: any): Promise<IInstruction[]> {
+        let instructons: any[] = recipeObject.analyzedInstructions[0].steps;
```
Comment (nashirj) [inline server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:162]: typo: instructions

**Review by nashirj — APPROVED** [review]

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts (hunk 5 above)
Comment (mplekunov) [reply server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:49]: #50 will be implemented in future branches...

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts (hunk 6 above)
Comment (mplekunov) [reply server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:52]: :(

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts (hunk 7 above)
Comment (mplekunov) [reply server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:66]: #50 will be implemented in future branches

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts (hunk 9 above)
Comment (mplekunov) [reply server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:69]: :(

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts (hunk 11 above)
Comment (mplekunov) [reply server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:67]: there is none. I don't include them if they are optional

File: server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts (hunk 12 above)
Comment (mplekunov) [reply server/recipeAPI/spoonacularAPI/SpoonacularRecipeAPI.ts:107]: #50 will be implemented in future branches

**Pull request merged**